### DISCUSSION TITLE
2 SysEx arrays to potentiometer

### DISCUSSION THREAD

**Discussion by RiNoize** [FortySevenEffects/arduino_midi_library#249] [Q&A] [answered]
Hi folks, In my Korg Kross 2  many parameters that can be assigned to a potentiometer use 2 SysEx arrays, a SysEx array on negative values and another SysEx array on positive values.
Its value ranges from -99 to +99
The center of the potentiometer travel is +00.

-99 = {0xf0, 0x42, 0x30, 0x00, 0x01, 0x49, 0x41, 0x00, 0x00, 0x09, 0x00, 0x34, 0x0f, 0x7f, 0x7f, 0x7f, 0x1d, 0xf7};

-01 = {0xf0, 0x42, 0x30, 0x00, 0x01, 0x49, 0x41, 0x00, 0x00, 0x09, 0x00, 0x34, 0x0f, 0x7f, 0x7f, 0x7f, 0x7f, 0xf7};

+00 = {0xf0, 0x42, 0x30, 0x00, 0x01, 0x49, 0x41, 0x00, 0x00, 0x09, 0x00, 0x34, 0x00, 0x00, 0x00, 0x00, 0x00, 0xf7};

+99 = {0xf0, 0x42, 0x30, 0x00, 0x01, 0x49, 0x41, 0x00, 0x00, 0x09, 0x00, 0x34, 0x00, 0x00, 0x00, 0x00, 0x63, 0xf7};

Does anyone know how to implement this in the sketch?
Thank you in advance.

**Comment by franky47** [comment, marked as answer]
From what I can see, the only thing that changes in those messages is the last 5 bytes of data (ignoring the last 0xf7 byte that marks the end of the SysEx message).

Those 5 bytes represent the value, and there is actually only one message: negative values in binary are represented using [two's complement](https://en.m.wikipedia.org/wiki/Two's_complement).

Now in the case of SysEx, there's a little extra trick: all data bytes have to be between 0x00 and 0x7f. That's because values with a most significant bit (MSB) of 1 (0x80 to 0xff) represent status codes and other control commands in the MIDI protocol.

From what we can see in the encoding of the -1 value (0x0f 0x7f 0x7f 0x7f 0x7f), we have 32 bits of data: that's a four byte value, spread out on four blocks of 7 bits (0x7f), plus one block of 4 bits (0x0f):

```
MIDI hex value          0x0f     0x7f     0x7f     0x7f     0x7f
Binary representation   00001111 01111111 01111111 01111111 01111111
Remove MSB              _0001111 _1111111 _1111111 _1111111 _1111111
Concatenate binary      00011111111111111111111111111111111
Group into bytes        11111111 11111111 11111111 11111111
Hex representation      0xffffffff
Two's complement        -1
```

 This is confirmed with the -99 value:
```
MIDI hex value          0x0f     0x7f     0x7f     0x7f     0x1d
Binary representation   00001111 01111111 01111111 01111111 00011101
Remove MSB               0001111  1111111  1111111  1111111  0011101
Concatenate binary      00011111111111111111111111110011101
Group into bytes        11111111 11111111 11111111 10011101
Hex representation      0xffffff9d
Two's complement        -99
```

So, how can we convert a potentiometer value into valid SysEx data?

The ADCs in Arduino have 10 bits of resolution (0 to 1023). It's more resolution than we need in this case (-99 to 99, plus zero, makes 199 values). We can use the `map` function to convert the analog reading into a -99/+99 ranged value.

If we store this -99 to 99 value directly on a signed 32 bit variable*, we can then perform our bit arithmetic to turn that into SysEx-compatible data:

```c++
const long rangedValue = map(analogRead(potPin), 0, 1023, -99, 99);

const byte sysExValueData[5] = { 
  (rangedValue >> 28) & 0x0f, // First 4 bits
  (rangedValue >> 21) & 0x7f, // Next 7 bits
  (rangedValue >> 14) & 0x7f, // etc..
  (rangedValue >> 7)  & 0x7f, // until
   rangedValue        & 0x7f  // the last 7 bits
};

const byte sysExMessage[16] = {
    0x42,
    0x30,
    0x00,
    0x01,
    0x49,
    0x41,
    0x00,
    0x00,
    0x09,
    0x00,
    0x34,
    sysExValueData[0],
    sysExValueData[1],
    sysExValueData[2],
    sysExValueData[3],
    sysExValueData[4],
};

MIDI.sendSysEx(16, sysExMessage); // The library sends start/end bytes by default
```

_* Note: using an `int` here may not work, as it's only [16 bits on AVR Arduinos](https://www.arduino.cc/reference/en/language/variables/data-types/int/). While there's plenty of space in a 16 bit variable to store -99/+99 values, it's two's complement has to cover the whole available space, so we use a `long` to make sure negative values are coded over 32 bits._

**Reply by RiNoize** [reply]
I had to ask a friend for help to implement your code in my sketch, after 4 hours we were able to implement it perfectly, now I have banks of 8 potentiometers, ready to run different arrays, Thank you Franky, my friend amazed with your code, from how you move the bits.
